File: apps/web/src/lib/api-client/tickets.ts
import { apiRequestJson, pathSegment } from "./client";
import type { components } from "./generated/openapi";

type ApiSchemas = components["schemas"];
type IntakeDetails = ApiSchemas["IntakeDetailsResponse"];

type ConversationStatus = "open" | "close_offered" | "closed";

type ChatMessage = {
  id: string;
  author: "user" | "assistant";
  body: string;
  createdAt: string;
};

type AttachmentMetadata = {
  id: string;
  name: string;
  description: string;
  sourceType: string;
  createdAt: string;
};

type AgentRun = {
  id: string;
  agentName: string;
  status: string;
  summary: string;
  safetyFlags: string[];
  createdAt: string;
};

type TimelineEntry = {
  id: string;
  eventType: string;
  body: string;
  actorUserId: string;
  createdAt: string;
};

type ClarificationRequest = {
  id: string;
  route: string;
  reason: string;
  questions: string[];
  createdAt: string;
};

type TicketCollaborator = {
  userId: string;
  username: string;
  displayName: string;
  access: "editor" | "viewer";
  addedByUserId: string;
  createdAt: string;
};

export type DirectoryUser = {
  id: string;
  username: string;
  displayName: string;
};

export type Ticket = Omit<
  ApiSchemas["TicketResponse"],
  | "state"
  | "conversationStatus"
  | "collectDisposition"
  | "collaborators"
  | "messages"
  | "attachments"
  | "agentRuns"
  | "clarificationRequests"
  | "timeline"
  | "customerStatus"
> & {
  state: TicketState;
  conversationStatus: ConversationStatus;
  collectDisposition: "raw" | "analysed" | null;
  collaborators: TicketCollaborator[];
  messages: ChatMessage[];
  attachments: AttachmentMetadata[];
  agentRuns: AgentRun[];
  clarificationRequests?: ClarificationRequest[];
  timeline: TimelineEntry[];
  customerStatus?: ApiSchemas["CustomerStatusResponse"];
};

export type TicketSummary = Omit<
  ApiSchemas["TicketSummaryResponse"],
  "state" | "customerStatus"
> & {
  state: TicketState;
  customerStatus?: ApiSchemas["CustomerStatusResponse"];
};

export type TicketState =
  | "DRAFT_INTAKE"
  | "INFO_REQUIRED"
  | "RFI_SEARCHING"
  | "RFI_SEARCH_INCOMPLETE"
  | "RFI_MATCH_OFFERED"
  | "ACTIVE_WORK_REVIEW"
  | "ACTIVE_WORK_SEARCH_INCOMPLETE"
  | "RFI_NO_MATCH"
  | "NEW_TASKING_CONSENT"
  | "JIOC_ROUTING_PENDING"
  | "JIOC_INTERVENTION_HOLD"
  | "JIOC_REVIEW"
  | "COLLECT_CHOICE"
  | "ANALYST_ASSIGNMENT"
  | "ANALYST_IN_PROGRESS"
  | "QC_REVIEW"
  | "REWORK_REQUIRED"
  | "MANAGER_APPROVAL"
  | "DISSEMINATION_READY"
  | "MANAGER_REANALYSIS_REVIEW"
  | "JIOC_REANALYSIS_ADJUDICATION"
  | "CLOSED_DELIVERED"
  | "CLOSED_REQUIREMENT_MET"
  | "CLOSED_REANALYSIS_DECLINED"
  | "CLOSED_EXISTING_PRODUCT_ACCEPTED"
  | "CLOSED_UNANSWERED"
  | "CLOSED_JOINED_EXISTING_WORK"
  | "CANCELLED";

export type IntakeUpdate = Partial<
  Pick<
    IntakeDetails,
    | "title"
    | "description"
    | "operationalQuestion"
    | "areaOrRegion"
    | "timePeriodStart"
    | "timePeriodEnd"
    | "priority"
    | "deadline"
    | "requiredOutputFormat"
    | "knownContext"
    | "restrictionsOrCaveats"
    | "customerSuccessCriteria"
    | "suggestedAcgContext"
    | "requestingUnit"
    | "intelligenceDisciplines"
    | "supportedOperation"
    | "urgencyJustification"
  >
>;

export type AttachmentMetadataInput = {
  name: string;
  description: string;
  sourceType: string;
};

export type TicketSummaryPage = {
  tickets: TicketSummary[];
  nextCursor: string | null;
};

export async function listTickets(cursor?: string): Promise<TicketSummaryPage> {
  const query = cursor ? `?cursor=${encodeURIComponent(cursor)}` : "";
  const response = await apiRequestJson<{ tickets: TicketSummary[]; nextCursor: string | null }>(
    `/api/v1/tickets${query}`,
    {
      method: "GET",
    },
  );
  return response;
}

export async function getTicket(ticketId: string): Promise<Ticket> {
  const response = await apiRequestJson<Ticket | { tickets: Ticket[] }>(
    `/api/v1/tickets/${pathSegment(ticketId)}`,
    {
      method: "GET",
    },
  );
  if ("tickets" in response) {
    const ticket = response.tickets.find((item) => item.id === ticketId);
    if (!ticket) {
      throw new Error("Ticket detail response did not contain the requested ticket.");
    }
    return ticket;
  }
  if (response.id !== ticketId) {
    throw new Error("Ticket detail response did not match the requested ticket.");
  }
  return response;
}

export async function sendChatMessage(
  payload: { ticketId?: string; message: string },
  csrfToken: string,
): Promise<Ticket> {
  return apiRequestJson<Ticket>("/api/v1/chat/messages", {
    body: JSON.stringify(payload),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function reopenTicketConversation(
  ticketId: string,
  csrfToken: string,
): Promise<Ticket> {
  return apiRequestJson<Ticket>(`/api/v1/tickets/${pathSegment(ticketId)}/conversation/reopen`, {
    headers: { "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function updateTicketIntake(
  ticketId: string,
  payload: IntakeUpdate,
  csrfToken: string,
): Promise<Ticket> {
  return apiRequestJson<Ticket>(`/api/v1/tickets/${pathSegment(ticketId)}/intake`, {
    body: JSON.stringify(payload),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "PATCH",
  });
}

export async function addTicketAttachment(
  ticketId: string,
  payload: AttachmentMetadataInput,
  csrfToken: string,
): Promise<Ticket> {
  return apiRequestJson<Ticket>(`/api/v1/tickets/${pathSegment(ticketId)}/attachments`, {
    body: JSON.stringify(payload),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function submitTicket(ticketId: string, csrfToken: string): Promise<Ticket> {
  return apiRequestJson<Ticket>(`/api/v1/tickets/${pathSegment(ticketId)}/submit`, {
    headers: { "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function cancelTicket(
  ticketId: string,
  reason: string,
  csrfToken: string,
): Promise<Ticket> {
  return apiRequestJson<Ticket>(`/api/v1/tickets/${pathSegment(ticketId)}/cancel`, {
    body: JSON.stringify({ reason }),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function consentNoMatch(
  ticketId: string,
  taskAsNewRequest: boolean,
  csrfToken: string,
): Promise<Ticket> {
  return apiRequestJson<Ticket>(`/api/v1/tickets/${pathSegment(ticketId)}/no-match-consent`, {
    body: JSON.stringify({ taskAsNewRequest }),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function chooseCollectOption(
  ticketId: string,
  analysed: boolean,
  csrfToken: string,
): Promise<Ticket> {
  return apiRequestJson<Ticket>(`/api/v1/tickets/${pathSegment(ticketId)}/collect-choice`, {
    body: JSON.stringify({ analysed }),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function addTicketInformation(
  ticketId: string,
  body: string,
  csrfToken: string,
): Promise<Ticket> {
  return apiRequestJson<Ticket>(`/api/v1/tickets/${pathSegment(ticketId)}/timeline`, {
    body: JSON.stringify({ body }),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function listUserDirectory(query: string): Promise<DirectoryUser[]> {
  const params = new URLSearchParams({ q: query });
  const response = await apiRequestJson<{ users: DirectoryUser[] }>(
    `/api/v1/users/directory?${params.toString()}`,
    { method: "GET" },
  );
  return response.users;
}

export async function decideProductOutcome(
  ticketId: string,
  payload: { meetsRequirement: boolean; reason: string; unmetCriteria: string[] },
  csrfToken: string,
): Promise<Ticket> {
  return apiRequestJson<Ticket>(`/api/v1/tickets/${pathSegment(ticketId)}/requirement-decision`, {
    body: JSON.stringify(payload),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function addTicketCollaborator(
  ticketId: string,
  username: string,
  access: "editor" | "viewer",
  csrfToken: string,
): Promise<Ticket> {
  return apiRequestJson<Ticket>(`/api/v1/tickets/${pathSegment(ticketId)}/collaborators`, {
    body: JSON.stringify({ username, access }),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function removeTicketCollaborator(
  ticketId: string,
  userId: string,
  csrfToken: string,
): Promise<Ticket> {
  return apiRequestJson<Ticket>(
    `/api/v1/tickets/${pathSegment(ticketId)}/collaborators/${pathSegment(userId)}`,
    {
      headers: { "X-CSRF-Token": csrfToken },
      method: "DELETE",
    },
  );
}

File: apps/api/src/coeus/schemas/tickets.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ChatMessageRequest(BaseModel):
    ticket_id: UUID | None = Field(default=None, validation_alias="ticketId")
    message: str = Field(min_length=3, max_length=4_000)


class IntakeUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=3, max_length=180)
    description: str | None = Field(default=None, min_length=3, max_length=2_000)
    operational_question: str | None = Field(
        default=None, min_length=3, max_length=500, validation_alias="operationalQuestion"
    )
    area_or_region: str | None = Field(
        default=None, min_length=2, max_length=180, validation_alias="areaOrRegion"
    )
    time_period_start: str | None = Field(
        default=None, max_length=80, validation_alias="timePeriodStart"
    )
    time_period_end: str | None = Field(
        default=None, max_length=80, validation_alias="timePeriodEnd"
    )
    priority: str | None = Field(default=None, min_length=2, max_length=40)
    deadline: str | None = Field(default=None, max_length=80)
    required_output_format: str | None = Field(
        default=None, min_length=2, max_length=120, validation_alias="requiredOutputFormat"
    )
    known_context: str | None = Field(
        default=None, max_length=2_000, validation_alias="knownContext"
    )
    restrictions_or_caveats: str | None = Field(
        default=None, max_length=1_000, validation_alias="restrictionsOrCaveats"
    )
    customer_success_criteria: str | None = Field(
        default=None, min_length=3, max_length=1_000, validation_alias="customerSuccessCriteria"
    )
    suggested_acg_context: str | None = Field(
        default=None, max_length=500, validation_alias="suggestedAcgContext"
    )
    requesting_unit: str | None = Field(
        default=None, min_length=2, max_length=180, validation_alias="requestingUnit"
    )
    intelligence_disciplines: str | None = Field(
        default=None, min_length=2, max_length=240, validation_alias="intelligenceDisciplines"
    )
    supported_operation: str | None = Field(
        default=None, min_length=2, max_length=180, validation_alias="supportedOperation"
    )
    urgency_justification: str | None = Field(
        default=None, min_length=3, max_length=500, validation_alias="urgencyJustification"
    )


class AttachmentMetadataRequest(BaseModel):
    name: str = Field(min_length=3, max_length=180)
    description: str = Field(min_length=3, max_length=500)
    source_type: str = Field(min_length=3, max_length=80, validation_alias="sourceType")


class AddInformationRequest(BaseModel):
    body: str = Field(min_length=3, max_length=2_000)


class CollaboratorAddRequest(BaseModel):
    username: str = Field(min_length=3, max_length=254)
    access: str = Field(pattern="^(editor|viewer)$")


class CollaboratorResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    user_id: UUID = Field(serialization_alias="userId")
    username: str
    display_name: str = Field(serialization_alias="displayName")
    access: str
    added_by_user_id: UUID = Field(serialization_alias="addedByUserId")
    created_at: datetime = Field(serialization_alias="createdAt")


class DirectoryUserResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    user_id: UUID = Field(serialization_alias="id")
    username: str
    display_name: str = Field(serialization_alias="displayName")


class DirectoryResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    users: list[DirectoryUserResponse]


class IntakeDetailsResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    title: str | None
    description: str | None
    operational_question: str | None = Field(serialization_alias="operationalQuestion")
    area_or_region: str | None = Field(serialization_alias="areaOrRegion")
    time_period_start: str | None = Field(serialization_alias="timePeriodStart")
    time_period_end: str | None = Field(serialization_alias="timePeriodEnd")
    priority: str | None
    deadline: str | None
    required_output_format: str | None = Field(serialization_alias="requiredOutputFormat")
    known_context: str | None = Field(serialization_alias="knownContext")
    restrictions_or_caveats: str | None = Field(serialization_alias="restrictionsOrCaveats")
    customer_success_criteria: str | None = Field(serialization_alias="customerSuccessCriteria")
    suggested_acg_context: str | None = Field(serialization_alias="suggestedAcgContext")
    requesting_unit: str | None = Field(serialization_alias="requestingUnit")
    intelligence_disciplines: str | None = Field(serialization_alias="intelligenceDisciplines")
    supported_operation: str | None = Field(serialization_alias="supportedOperation")
    urgency_justification: str | None = Field(serialization_alias="urgencyJustification")
    missing_information: list[str] = Field(serialization_alias="missingInformation")
    confidence: float


class IntakeChecklistItemResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    key: str
    label: str
    value: str | None
    satisfied: bool


class ChatMessageResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    message_id: UUID = Field(serialization_alias="id")
    author: str
    body: str
    created_at: datetime = Field(serialization_alias="createdAt")


class AttachmentMetadataResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    attachment_id: UUID = Field(serialization_alias="id")
    name: str
    description: str
    source_type: str = Field(serialization_alias="sourceType")
    created_at: datetime = Field(serialization_alias="createdAt")


class AgentRunResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    run_id: UUID = Field(serialization_alias="id")
    agent_name: str = Field(serialization_alias="agentName")
    status: str
    summary: str
    safety_flags: list[str] = Field(serialization_alias="safetyFlags")
    created_at: datetime = Field(serialization_alias="createdAt")


class TimelineEntryResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    entry_id: UUID = Field(serialization_alias="id")
    event_type: str = Field(serialization_alias="eventType")
    body: str
    actor_user_id: UUID = Field(serialization_alias="actorUserId")
    created_at: datetime = Field(serialization_alias="createdAt")


class ClarificationResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    clarification_id: UUID = Field(serialization_alias="id")
    route: str
    reason: str
    questions: list[str]
    created_at: datetime = Field(serialization_alias="createdAt")


class TicketCancelRequest(BaseModel):
    reason: str = Field(min_length=3, max_length=300)


class NoMatchConsentRequest(BaseModel):
    task_as_new_request: bool = Field(validation_alias="taskAsNewRequest")


class CollectChoiceRequest(BaseModel):
    # True: collect followed by RFA analysis; False: raw collect only.
    analysed: bool


class CustomerEstimateResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    earliest: datetime | None
    likely: datetime | None
    latest: datetime | None
    confidence: str
    status: str
    as_of: datetime = Field(serialization_alias="asOf")
    policy_version: str = Field(serialization_alias="policyVersion")


class CustomerJourneyStageResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    code: str
    label: str
    status: str


class CustomerStatusResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    code: str
    label: str
    explanation: str
    current_leg: str = Field(serialization_alias="currentLeg")
    action_required: bool = Field(serialization_alias="actionRequired")
    action_type: str | None = Field(serialization_alias="actionType")
    next_milestone: str | None = Field(serialization_alias="nextMilestone")
    canonical_ticket_id: UUID | None = Field(serialization_alias="canonicalTicketId")
    estimate: CustomerEstimateResponse | None
    journey: list[CustomerJourneyStageResponse]


class TicketResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    ticket_id: UUID = Field(serialization_alias="id")
    reference: str
    requester_user_id: UUID = Field(serialization_alias="requesterUserId")
    state: str
    customer_status: CustomerStatusResponse = Field(serialization_alias="customerStatus")
    intake: IntakeDetailsResponse
    intake_checklist: list[IntakeChecklistItemResponse] = Field(
        serialization_alias="intakeChecklist"
    )
    conversation_status: str = Field(serialization_alias="conversationStatus")
    collect_disposition: str | None = Field(serialization_alias="collectDisposition")
    is_ready_for_submission: bool = Field(serialization_alias="isReadyForSubmission")
    visible_product_matches: list[str] = Field(serialization_alias="visibleProductMatches")
    released_product_ids: list[UUID] = Field(serialization_alias="releasedProductIds")
    collaborators: list[CollaboratorResponse]
    messages: list[ChatMessageResponse]
    attachments: list[AttachmentMetadataResponse]
    agent_runs: list[AgentRunResponse] = Field(serialization_alias="agentRuns")
    clarification_requests: list[ClarificationResponse] = Field(
        serialization_alias="clarificationRequests"
    )
    timeline: list[TimelineEntryResponse]
    created_at: datetime = Field(serialization_alias="createdAt")
    updated_at: datetime = Field(serialization_alias="updatedAt")


class TicketSummaryResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    ticket_id: UUID = Field(serialization_alias="id")
    reference: str
    requester_user_id: UUID = Field(serialization_alias="requesterUserId")
    state: str
    customer_status: CustomerStatusResponse = Field(serialization_alias="customerStatus")
    title: str | None
    priority: str | None
    is_ready_for_submission: bool = Field(serialization_alias="isReadyForSubmission")
    collaborator_count: int = Field(serialization_alias="collaboratorCount")
    released_product_id: UUID | None = Field(serialization_alias="releasedProductId")
    created_at: datetime = Field(serialization_alias="createdAt")
    updated_at: datetime = Field(serialization_alias="updatedAt")


class TicketListResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    tickets: list[TicketSummaryResponse]
    next_cursor: UUID | None = Field(serialization_alias="nextCursor")
